File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        List<List<Integer>> answer = new ArrayList<>();
        PriorityQueue<Integer> minq = new PriorityQueue<>();
        PriorityQueue<Integer> maxq = new PriorityQueue<>(Collections.reverseOrder());
        
        for (String operation : operations) {
            String[] part= operation.split(" ");
            String command = part[0];
            int number = Integer.parseInt(part[1]);
            
            if (command.equals("I")) {
                minq.offer(number);
                maxq.offer(number);
            } else if (command.equals("D")) {
                if (number == 1) {
                    if (maxq.isEmpty()) {
                        continue;
                    }
                    int max = maxq.poll();
                    minq.remove(max);
                } else {
                    if(minq.isEmpty()) {
                        continue;
                    }
                    int min = minq.poll();
                    maxq.remove(min);
                }
                
            } 
        }
        
        if (!minq.isEmpty() || !maxq.isEmpty()) {
            answer.add(Arrays.asList(maxq.peek(), minq.peek()));

        }
        answer.add(Arrays.asList(0,0));

        return answer.get(0).stream().mapToInt(i -> i).toArray();
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
from heapq import heappush, heappop

def solution(operations):
    min_q = []  
    max_q = []  

    for op in operations:
        command, num_str = op.split()
        num = int(num_str)

        if command == 'I':
            heappush(min_q, num)
            heappush(max_q, -num)
        elif command == 'D':
            if num == 1 and max_q:
                max_val = -heappop(max_q)
                min_q.remove(max_val)
            elif num == -1 and min_q: 
                min_val = heappop(min_q)
                max_q.remove(-min_val)

    if not min_q:
        return [0, 0]

    return [-heappop(max_q), heappop(min_q)]


